File: packages/snow_draw_core/lib/draw/elements/types/arrow/elbow/elbow_geometry.dart
import '../../../../models/element_state.dart';
import '../../../../types/draw_point.dart';
import '../../../../types/draw_rect.dart';
import '../../../../utils/selection_calculator.dart';
import '../arrow_binding.dart';
import 'elbow_constants.dart';
import 'elbow_heading.dart';

/// Internal axis tagging for elbow paths.
enum ElbowAxis {
  horizontal,
  vertical;

  /// Whether this axis is horizontal.
  bool get isHorizontal => this == ElbowAxis.horizontal;

  /// Whether this axis is vertical.
  bool get isVertical => this == ElbowAxis.vertical;
}

/// Shared geometry helpers for elbow routing and editing.
final class ElbowGeometry {
  const ElbowGeometry._();

  static const _headingEpsilon = 1e-6;

  /// Returns the dominant cardinal heading for a vector.
  ///
  /// The dominant axis wins; ties favor horizontal headings.
  static ElbowHeading headingForVector(double dx, double dy) {
    final absX = dx.abs();
    final absY = dy.abs();
    if (absX >= absY) {
      return dx >= 0 ? ElbowHeading.right : ElbowHeading.left;
    }
    return dy >= 0 ? ElbowHeading.down : ElbowHeading.up;
  }

  /// Returns the heading for a segment from [from] to [to].
  static ElbowHeading headingForSegment(DrawPoint from, DrawPoint to) =>
      headingForVector(to.x - from.x, to.y - from.y);

  /// Manhattan distance between two points.
  static double manhattanDistance(DrawPoint a, DrawPoint b) =>
      (a.x - b.x).abs() + (a.y - b.y).abs();

  /// Returns true when the segment is closer to horizontal than vertical.
  static bool isHorizontal(DrawPoint a, DrawPoint b) =>
      (a.y - b.y).abs() <= (a.x - b.x).abs();

  /// Determines which side of the bounds a point belongs to, using
  /// a scaled-triangle quadrant test around the center.
  static ElbowHeading headingForPointOnBounds(
    DrawRect bounds,
    DrawPoint point,
  ) {
    final center = bounds.center;
    const scale = 2.0;
    final topLeft = _scalePointFromOrigin(
      DrawPoint(x: bounds.minX, y: bounds.minY),
      center,
      scale,
    );
    final topRight = _scalePointFromOrigin(
      DrawPoint(x: bounds.maxX, y: bounds.minY),
      center,
      scale,
    );
    final bottomLeft = _scalePointFromOrigin(
      DrawPoint(x: bounds.minX, y: bounds.maxY),
      center,
      scale,
    );
    final bottomRight = _scalePointFromOrigin(
      DrawPoint(x: bounds.maxX, y: bounds.maxY),
      center,
      scale,
    );

    if (_triangleContainsPoint(topLeft, topRight, center, point)) {
      return ElbowHeading.up;
    }
    if (_triangleContainsPoint(topRight, bottomRight, center, point)) {
      return ElbowHeading.right;
    }
    if (_triangleContainsPoint(bottomRight, bottomLeft, center, point)) {
      return ElbowHeading.down;
    }
    return ElbowHeading.left;
  }

  static DrawPoint _scalePointFromOrigin(
    DrawPoint point,
    DrawPoint origin,
    double scale,
  ) => DrawPoint(
    x: origin.x + (point.x - origin.x) * scale,
    y: origin.y + (point.y - origin.y) * scale,
  );

  static DrawPoint _vectorFromPoints(DrawPoint to, DrawPoint from) =>
      DrawPoint(x: to.x - from.x, y: to.y - from.y);

  static double _dotProduct(DrawPoint a, DrawPoint b) => a.x * b.x + a.y * b.y;

  static bool _triangleContainsPoint(
    DrawPoint a,
    DrawPoint b,
    DrawPoint c,
    DrawPoint point,
  ) {
    final v0 = _vectorFromPoints(c, a);
    final v1 = _vectorFromPoints(b, a);
    final v2 = _vectorFromPoints(point, a);

    final dot00 = _dotProduct(v0, v0);
    final dot01 = _dotProduct(v0, v1);
    final dot02 = _dotProduct(v0, v2);
    final dot11 = _dotProduct(v1, v1);
    final dot12 = _dotProduct(v1, v2);

    final denom = dot00 * dot11 - dot01 * dot01;
    if (denom.abs() <= _headingEpsilon) {
      return false;
    }
    final invDenom = 1 / denom;
    final u = (dot11 * dot02 - dot01 * dot12) * invDenom;
    final v = (dot00 * dot12 - dot01 * dot02) * invDenom;

    return u >= -_headingEpsilon &&
        v >= -_headingEpsilon &&
        u + v <= 1 + _headingEpsilon;
  }

  // --- Methods absorbed from ElbowPathUtils ---

  /// Returns an axis only when the segment is axis-aligned within tolerance.
  static ElbowAxis? axisAlignedForSegment(
    DrawPoint a,
    DrawPoint b, {
    double epsilon = ElbowConstants.dedupThreshold,
  }) {
    final dx = (a.x - b.x).abs();
    final dy = (a.y - b.y).abs();
    if (dx <= epsilon && dy <= epsilon) {
      return null;
    }
    if (dy <= epsilon) {
      return ElbowAxis.horizontal;
    }
    if (dx <= epsilon) {
      return ElbowAxis.vertical;
    }
    return null;
  }

  /// Returns a stable axis for a segment, falling back to the dominant axis.
  static ElbowAxis axisForSegment(
    DrawPoint a,
    DrawPoint b, {
    double epsilon = ElbowConstants.dedupThreshold,
  }) {
    final aligned = axisAlignedForSegment(a, b, epsilon: epsilon);
    if (aligned != null) {
      return aligned;
    }
    return isHorizontal(a, b) ? ElbowAxis.horizontal : ElbowAxis.vertical;
  }

  /// Returns true when a segment is (or should be treated as) horizontal.
  static bool segmentIsHorizontal(
    DrawPoint a,
    DrawPoint b, {
    double epsilon = ElbowConstants.dedupThreshold,
  }) => axisForSegment(a, b, epsilon: epsilon).isHorizontal;

  /// Returns true when a segment is (or should be treated as) vertical.
  static bool segmentIsVertical(
    DrawPoint a,
    DrawPoint b, {
    double epsilon = ElbowConstants.dedupThreshold,
  }) => axisForSegment(a, b, epsilon: epsilon).isVertical;

  /// Returns the shared axis coordinate for a segment.
  static double axisValue(
    DrawPoint start,
    DrawPoint end, {
    required ElbowAxis axis,
  }) => axis.isHorizontal ? (start.y + end.y) / 2 : (start.x + end.x) / 2;

  /// Returns true when two points are nearly identical.
  static bool pointsClose(
    DrawPoint a,
    DrawPoint b, {
    double epsilon = ElbowConstants.dedupThreshold,
  }) => (a.x - b.x).abs() <= epsilon && (a.y - b.y).abs() <= epsilon;

  /// Returns true when two points align on either axis.
  static bool pointsAligned(
    DrawPoint a,
    DrawPoint b, {
    double epsilon = ElbowConstants.dedupThreshold,
  }) => (a.x - b.x).abs() <= epsilon || (a.y - b.y).abs() <= epsilon;

  /// Returns true when three points form a straight orthogonal line.
  static bool segmentsCollinear(
    DrawPoint a,
    DrawPoint b,
    DrawPoint c, {
    double epsilon = ElbowConstants.dedupThreshold,
  }) {
    final axis = axisAlignedForSegment(a, b, epsilon: epsilon);
    final nextAxis = axisAlignedForSegment(b, c, epsilon: epsilon);
    if (axis == null || nextAxis == null || axis != nextAxis) {
      return false;
    }
    final axisValueA = axisValue(a, b, axis: axis);
    final axisValueB = axisValue(b, c, axis: axis);
    return (axisValueA - axisValueB).abs() <= epsilon;
  }

  /// Returns a direct elbow path with a single corner when possible.
  static List<DrawPoint> directElbowPath(
    DrawPoint start,
    DrawPoint end, {
    required bool preferHorizontal,
    double epsilon = ElbowConstants.dedupThreshold,
  }) {
    if ((start.x - end.x).abs() <= epsilon ||
        (start.y - end.y).abs() <= epsilon) {
      return [start, end];
    }
    final mid = preferHorizontal
        ? DrawPoint(x: end.x, y: start.y)
        : DrawPoint(x: start.x, y: end.y);
    if (mid == start || mid == end) {
      return [start, end];
    }
    return [start, mid, end];
  }

  /// Removes short interior segments while keeping endpoints intact.
  static List<DrawPoint> removeShortSegments(
    List<DrawPoint> points, {
    double minLength = ElbowConstants.dedupThreshold,
  }) {
    if (points.length < 4) {
      return points;
    }
    return [
      points.first,
      for (var i = 1; i < points.length - 1; i++)
        if (manhattanDistance(points[i - 1], points[i]) > minLength) points[i],
      points.last,
    ];
  }

  /// Keeps only the corner points of an orthogonal polyline.
  static List<DrawPoint> cornerPoints(List<DrawPoint> points) {
    if (points.length <= 2) {
      return points;
    }
    var prevH = segmentIsHorizontal(points[0], points[1]);
    final result = <DrawPoint>[points.first];
    for (var i = 1; i < points.length - 1; i++) {
      final nextH = segmentIsHorizontal(points[i], points[i + 1]);
      if (prevH != nextH) {
        result.add(points[i]);
      }
      prevH = nextH;
    }
    result.add(points.last);
    return result;
  }

  /// Simplifies an orthogonal path while keeping pinned points intact.
  static List<DrawPoint> simplifyPath(
    List<DrawPoint> points, {
    Set<DrawPoint> pinned = const <DrawPoint>{},
  }) {
    if (points.length < 3) {
      return points;
    }

    // Remove collinear interior points (keep pinned).
    final reduced = <DrawPoint>[points.first];
    for (var i = 1; i < points.length - 1; i++) {
      final p = points[i];
      if (pinned.contains(p) ||
          segmentIsHorizontal(reduced.last, p) !=
              segmentIsHorizontal(p, points[i + 1])) {
        reduced.add(p);
      }
    }
    reduced.add(points.last);

    // Remove near-duplicate consecutive points.
    final cleaned = <DrawPoint>[reduced.first];
    for (var i = 1; i < reduced.length; i++) {
      final p = reduced[i];
      if (p != cleaned.last &&
          (pinned.contains(p) ||
              manhattanDistance(cleaned.last, p) >
                  ElbowConstants.dedupThreshold)) {
        cleaned.add(p);
      }
    }
    return List<DrawPoint>.unmodifiable(cleaned);
  }

  /// Merges consecutive segments that share the same heading.
  ///
  /// Two segments with the same heading (e.g. both Right) but at
  /// different axis values indicate a redundant intermediate point.
  /// Removing it collapses the pair into a single segment.
  /// Pinned points are preserved.
  static List<DrawPoint> mergeConsecutiveSameHeading(
    List<DrawPoint> points, {
    Set<DrawPoint> pinned = const <DrawPoint>{},
  }) {
    if (points.length < 3) {
      return points;
    }
    var changed = true;
    var current = points;
    while (changed) {
      changed = false;
      final result = <DrawPoint>[current.first];
      for (var i = 1; i < current.length - 1; i++) {
        final prev = result.last;
        final mid = current[i];
        final next = current[i + 1];
        final prevLen = manhattanDistance(prev, mid);
        final nextLen = manhattanDistance(mid, next);
        if (prevLen <= ElbowConstants.dedupThreshold ||
            nextLen <= ElbowConstants.dedupThreshold) {
          result.add(mid);
          continue;
        }
        final prevH = headingForSegment(prev, mid);
        final nextH = headingForSegment(mid, next);
        if (prevH == nextH && !pinned.contains(mid)) {
          changed = true;
          continue;
        }
        result.add(mid);
      }
      result.add(current.last);
      current = result;
    }
    return List<DrawPoint>.unmodifiable(current);
  }

  /// Returns true when any segment is diagonal beyond the tolerance.
  static bool hasDiagonalSegments(List<DrawPoint> points) {
    for (var i = 1; i < points.length; i++) {
      if ((points[i].x - points[i - 1].x).abs() >
              ElbowConstants.dedupThreshold &&
          (points[i].y - points[i - 1].y).abs() >
              ElbowConstants.dedupThreshold) {
        return true;
      }
    }
    return false;
  }

  /// Offsets [point] along [heading] by [distance].
  static DrawPoint offsetPoint(
    DrawPoint point,
    ElbowHeading heading,
    double distance,
  ) => DrawPoint(
    x: point.x + heading.dx * distance,
    y: point.y + heading.dy * distance,
  );

  /// Total Manhattan path length across all segments.
  static double pathLength(List<DrawPoint> points) {
    var length = 0.0;
    for (var i = 0; i < points.length - 1; i++) {
      length += manhattanDistance(points[i], points[i + 1]);
    }
    return length;
  }

  /// Whether two point lists are element-wise equal.
  static bool pointListsEqual(List<DrawPoint> a, List<DrawPoint> b) {
    if (a.length != b.length) {
      return false;
    }
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) {
        return false;
      }
    }
    return true;
  }

  /// Whether two point lists are equal except at the first and last
  /// positions.
  static bool pointListsEqualExceptEndpoints(
    List<DrawPoint> a,
    List<DrawPoint> b,
  ) {
    if (a.length != b.length || a.length < 2) {
      return false;
    }
    for (var i = 1; i < a.length - 1; i++) {
      if (a[i] != b[i]) {
        return false;
      }
    }
    return true;
  }

  /// Resolves the bound heading for an arrow endpoint.
  ///
  /// Returns the cardinal direction the arrow should exit from the
  /// bound element, or `null` when the binding target is missing.
  static ElbowHeading? resolveBoundHeading({
    required ArrowBinding binding,
    required Map<String, ElementState> elementsById,
    required DrawPoint point,
  }) {
    final element = elementsById[binding.elementId];
    if (element == null) {
      return null;
    }
    final bounds = SelectionCalculator.computeElementWorldAabb(element);
    final anchor = ArrowBindingUtils.resolveElbowAnchorPoint(
      binding: binding,
      target: element,
    );
    return headingForPointOnBounds(bounds, anchor ?? point);
  }
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/elbow/elbow_geometry.rs
#![allow(dead_code)]

use std::collections::{HashMap, HashSet};

use crate::draw::elements::types::arrow::arrow_binding::{
    ArrowBinding, ArrowBindingUtils, ElementState as ArrowElementState,
};
use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;

use super::elbow_constants::ElbowConstants;

/// Internal axis tagging for elbow paths.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ElbowAxis {
    Horizontal,
    Vertical,
}

impl ElbowAxis {
    /// Whether this axis is horizontal.
    pub const fn is_horizontal(self) -> bool {
        matches!(self, Self::Horizontal)
    }

    /// Whether this axis is vertical.
    pub const fn is_vertical(self) -> bool {
        matches!(self, Self::Vertical)
    }
}

/// Cardinal heading used by elbow path geometry.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ElbowHeading {
    Up,
    Right,
    Down,
    Left,
}

impl ElbowHeading {
    /// Horizontal unit component of this heading.
    pub const fn dx(self) -> f64 {
        match self {
            Self::Left => -1.0,
            Self::Right => 1.0,
            _ => 0.0,
        }
    }

    /// Vertical unit component of this heading.
    pub const fn dy(self) -> f64 {
        match self {
            Self::Up => -1.0,
            Self::Down => 1.0,
            _ => 0.0,
        }
    }
}

/// Shared geometry helpers for elbow routing and editing.
pub struct ElbowGeometry;

impl ElbowGeometry {
    const HEADING_EPSILON: f64 = 1e-6;

    /// Returns the dominant cardinal heading for a vector.
    ///
    /// The dominant axis wins; ties favor horizontal headings.
    pub fn heading_for_vector(dx: f64, dy: f64) -> ElbowHeading {
        let abs_x = dx.abs();
        let abs_y = dy.abs();
        if abs_x >= abs_y {
            if dx >= 0.0 {
                ElbowHeading::Right
            } else {
                ElbowHeading::Left
            }
        } else if dy >= 0.0 {
            ElbowHeading::Down
        } else {
            ElbowHeading::Up
        }
    }

    /// Returns the heading for a segment from `from` to `to`.
    pub fn heading_for_segment(from: DrawPoint, to: DrawPoint) -> ElbowHeading {
        Self::heading_for_vector(to.x - from.x, to.y - from.y)
    }

    /// Manhattan distance between two points.
    pub fn manhattan_distance(a: DrawPoint, b: DrawPoint) -> f64 {
        (a.x - b.x).abs() + (a.y - b.y).abs()
    }

    /// Returns true when the segment is closer to horizontal than vertical.
    pub fn is_horizontal(a: DrawPoint, b: DrawPoint) -> bool {
        (a.y - b.y).abs() <= (a.x - b.x).abs()
    }

    /// Determines which side of the bounds a point belongs to.
    pub fn heading_for_point_on_bounds(bounds: DrawRect, point: DrawPoint) -> ElbowHeading {
        let center = bounds.center();
        let dx = point.x - center.x;
        let dy = point.y - center.y;
        let width = bounds.width().abs();
        let height = bounds.height().abs();
        if width <= Self::HEADING_EPSILON || height <= Self::HEADING_EPSILON {
            return ElbowHeading::Left;
        }

        let horizontal_weight = dx.abs() * height;
        let vertical_weight = dy.abs() * width;
        let tolerance = Self::HEADING_EPSILON * width * height;

        if dy <= Self::HEADING_EPSILON
            && dy >= -height - Self::HEADING_EPSILON
            && horizontal_weight <= vertical_weight + tolerance
        {
            return ElbowHeading::Up;
        }
        if dx >= -Self::HEADING_EPSILON
            && dx <= width + Self::HEADING_EPSILON
            && vertical_weight <= horizontal_weight + tolerance
        {
            return ElbowHeading::Right;
        }
        if dy >= -Self::HEADING_EPSILON
            && dy <= height + Self::HEADING_EPSILON
            && horizontal_weight <= vertical_weight + tolerance
        {
            return ElbowHeading::Down;
        }

        ElbowHeading::Left
    }

    /// Returns an axis only when the segment is axis-aligned within tolerance.
    pub fn axis_aligned_for_segment(a: DrawPoint, b: DrawPoint) -> Option<ElbowAxis> {
        Self::axis_aligned_for_segment_with_epsilon(a, b, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Returns an axis only when the segment is axis-aligned within tolerance.
    pub fn axis_aligned_for_segment_with_epsilon(
        a: DrawPoint,
        b: DrawPoint,
        epsilon: f64,
    ) -> Option<ElbowAxis> {
        let dx = (a.x - b.x).abs();
        let dy = (a.y - b.y).abs();
        if dx <= epsilon && dy <= epsilon {
            return None;
        }
        if dy <= epsilon {
            return Some(ElbowAxis::Horizontal);
        }
        if dx <= epsilon {
            return Some(ElbowAxis::Vertical);
        }
        None
    }

    /// Returns a stable axis for a segment, falling back to the dominant axis.
    pub fn axis_for_segment(a: DrawPoint, b: DrawPoint) -> ElbowAxis {
        Self::axis_for_segment_with_epsilon(a, b, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Returns a stable axis for a segment, falling back to the dominant axis.
    pub fn axis_for_segment_with_epsilon(a: DrawPoint, b: DrawPoint, epsilon: f64) -> ElbowAxis {
        Self::axis_aligned_for_segment_with_epsilon(a, b, epsilon).unwrap_or_else(|| {
            if Self::is_horizontal(a, b) {
                ElbowAxis::Horizontal
            } else {
                ElbowAxis::Vertical
            }
        })
    }

    /// Returns true when a segment is (or should be treated as) horizontal.
    pub fn segment_is_horizontal(a: DrawPoint, b: DrawPoint) -> bool {
        Self::segment_is_horizontal_with_epsilon(a, b, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Returns true when a segment is (or should be treated as) horizontal.
    pub fn segment_is_horizontal_with_epsilon(a: DrawPoint, b: DrawPoint, epsilon: f64) -> bool {
        Self::axis_for_segment_with_epsilon(a, b, epsilon).is_horizontal()
    }

    /// Returns true when a segment is (or should be treated as) vertical.
    pub fn segment_is_vertical(a: DrawPoint, b: DrawPoint) -> bool {
        Self::segment_is_vertical_with_epsilon(a, b, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Returns true when a segment is (or should be treated as) vertical.
    pub fn segment_is_vertical_with_epsilon(a: DrawPoint, b: DrawPoint, epsilon: f64) -> bool {
        Self::axis_for_segment_with_epsilon(a, b, epsilon).is_vertical()
    }

    /// Returns the shared axis coordinate for a segment.
    pub fn axis_value(start: DrawPoint, end: DrawPoint, axis: ElbowAxis) -> f64 {
        if axis.is_horizontal() {
            (start.y + end.y) / 2.0
        } else {
            (start.x + end.x) / 2.0
        }
    }

    /// Returns true when two points are nearly identical.
    pub fn points_close(a: DrawPoint, b: DrawPoint) -> bool {
        Self::points_close_with_epsilon(a, b, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Returns true when two points are nearly identical.
    pub fn points_close_with_epsilon(a: DrawPoint, b: DrawPoint, epsilon: f64) -> bool {
        (a.x - b.x).abs() <= epsilon && (a.y - b.y).abs() <= epsilon
    }

    /// Returns true when two points align on either axis.
    pub fn points_aligned(a: DrawPoint, b: DrawPoint) -> bool {
        Self::points_aligned_with_epsilon(a, b, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Returns true when two points align on either axis.
    pub fn points_aligned_with_epsilon(a: DrawPoint, b: DrawPoint, epsilon: f64) -> bool {
        (a.x - b.x).abs() <= epsilon || (a.y - b.y).abs() <= epsilon
    }

    /// Returns true when three points form a straight orthogonal line.
    pub fn segments_collinear(a: DrawPoint, b: DrawPoint, c: DrawPoint) -> bool {
        Self::segments_collinear_with_epsilon(a, b, c, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Returns true when three points form a straight orthogonal line.
    pub fn segments_collinear_with_epsilon(
        a: DrawPoint,
        b: DrawPoint,
        c: DrawPoint,
        epsilon: f64,
    ) -> bool {
        let Some(axis) = Self::axis_aligned_for_segment_with_epsilon(a, b, epsilon) else {
            return false;
        };
        let Some(next_axis) = Self::axis_aligned_for_segment_with_epsilon(b, c, epsilon) else {
            return false;
        };
        if axis != next_axis {
            return false;
        }

        let axis_value_a = Self::axis_value(a, b, axis);
        let axis_value_b = Self::axis_value(b, c, axis);
        (axis_value_a - axis_value_b).abs() <= epsilon
    }

    /// Returns a direct elbow path with a single corner when possible.
    pub fn direct_elbow_path(
        start: DrawPoint,
        end: DrawPoint,
        prefer_horizontal: bool,
    ) -> Vec<DrawPoint> {
        Self::direct_elbow_path_with_epsilon(
            start,
            end,
            prefer_horizontal,
            ElbowConstants::DEDUP_THRESHOLD,
        )
    }

    /// Returns a direct elbow path with a single corner when possible.
    pub fn direct_elbow_path_with_epsilon(
        start: DrawPoint,
        end: DrawPoint,
        prefer_horizontal: bool,
        epsilon: f64,
    ) -> Vec<DrawPoint> {
        if (start.x - end.x).abs() <= epsilon || (start.y - end.y).abs() <= epsilon {
            return vec![start, end];
        }

        let mid = if prefer_horizontal {
            DrawPoint::new(end.x, start.y)
        } else {
            DrawPoint::new(start.x, end.y)
        };
        vec![start, mid, end]
    }

    /// Removes short interior segments while keeping endpoints intact.
    pub fn remove_short_segments(points: &[DrawPoint]) -> Vec<DrawPoint> {
        Self::remove_short_segments_with_min_length(points, ElbowConstants::DEDUP_THRESHOLD)
    }

    /// Removes short interior segments while keeping endpoints intact.
    pub fn remove_short_segments_with_min_length(
        points: &[DrawPoint],
        min_length: f64,
    ) -> Vec<DrawPoint> {
        if points.len() < 4 {
            return points.to_vec();
        }

        let mut result = Vec::with_capacity(points.len());
        result.push(points[0]);
        for index in 1..(points.len() - 1) {
            if Self::manhattan_distance(points[index - 1], points[index]) > min_length {
                result.push(points[index]);
            }
        }
        result.push(points[points.len() - 1]);
        result
    }

    /// Keeps only the corner points of an orthogonal polyline.
    pub fn corner_points(points: &[DrawPoint]) -> Vec<DrawPoint> {
        if points.len() <= 2 {
            return points.to_vec();
        }

        let mut prev_horizontal = Self::segment_is_horizontal(points[0], points[1]);
        let mut result = Vec::with_capacity(points.len());
        result.push(points[0]);

        for index in 1..(points.len() - 1) {
            let next_horizontal = Self::segment_is_horizontal(points[index], points[index + 1]);
            if prev_horizontal != next_horizontal {
                result.push(points[index]);
            }
            prev_horizontal = next_horizontal;
        }

        result.push(points[points.len() - 1]);
        result
    }

    /// Simplifies an orthogonal path while keeping pinned points intact.
    pub fn simplify_path(points: &[DrawPoint]) -> Vec<DrawPoint> {
        let pinned = HashSet::new();
        Self::simplify_path_with_pinned(points, &pinned)
    }

    /// Simplifies an orthogonal path while keeping pinned points intact.
    pub fn simplify_path_with_pinned(
        points: &[DrawPoint],
        pinned: &HashSet<DrawPoint>,
    ) -> Vec<DrawPoint> {
        if points.len() < 3 {
            return points.to_vec();
        }

        let mut reduced = Vec::with_capacity(points.len());
        reduced.push(points[0]);
        for index in 1..(points.len() - 1) {
            let point = points[index];
            let last = *reduced.last().expect("reduced always has first point");
            let has_turn = Self::segment_is_horizontal(last, point)
                != Self::segment_is_horizontal(point, points[index + 1]);
            if pinned.contains(&point) || has_turn {
                reduced.push(point);
            }
        }
        reduced.push(points[points.len() - 1]);

        let mut cleaned = Vec::with_capacity(reduced.len());
        cleaned.push(reduced[0]);
        for point in reduced.iter().copied().skip(1) {
            let last = *cleaned.last().expect("cleaned always has first point");
            if point != last
                && (pinned.contains(&point)
                    || Self::manhattan_distance(last, point) > ElbowConstants::DEDUP_THRESHOLD)
            {
                cleaned.push(point);
            }
        }

        cleaned
    }

    /// Merges consecutive segments that share the same heading.
    pub fn merge_consecutive_same_heading(points: &[DrawPoint]) -> Vec<DrawPoint> {
        let pinned = HashSet::new();
        Self::merge_consecutive_same_heading_with_pinned(points, &pinned)
    }

    /// Merges consecutive segments that share the same heading.
    ///
    /// Two segments with the same heading but at different axis values indicate
    /// a redundant intermediate point. Removing it collapses the pair into a
    /// single segment.
    pub fn merge_consecutive_same_heading_with_pinned(
        points: &[DrawPoint],
        pinned: &HashSet<DrawPoint>,
    ) -> Vec<DrawPoint> {
        if points.len() < 3 {
            return points.to_vec();
        }

        let mut merged = Vec::with_capacity(points.len());
        merged.push(points[0]);
        merged.push(points[1]);

        for point in points.iter().copied().skip(2) {
            merged.push(point);
            while merged.len() >= 3 {
                let prev = merged[merged.len() - 3];
                let mid = merged[merged.len() - 2];
                let next = merged[merged.len() - 1];
                let prev_len = Self::manhattan_distance(prev, mid);
                let next_len = Self::manhattan_distance(mid, next);

                if prev_len <= ElbowConstants::DEDUP_THRESHOLD
                    || next_len <= ElbowConstants::DEDUP_THRESHOLD
                    || pinned.contains(&mid)
                    || Self::heading_for_segment(prev, mid) != Self::heading_for_segment(mid, next)
                {
                    break;
                }

                merged.remove(merged.len() - 2);
            }
        }

        merged
    }

    /// Returns true when any segment is diagonal beyond the tolerance.
    pub fn has_diagonal_segments(points: &[DrawPoint]) -> bool {
        for index in 1..points.len() {
            if (points[index].x - points[index - 1].x).abs() > ElbowConstants::DEDUP_THRESHOLD
                && (points[index].y - points[index - 1].y).abs() > ElbowConstants::DEDUP_THRESHOLD
            {
                return true;
            }
        }
        false
    }

    /// Offsets `point` along `heading` by `distance`.
    pub fn offset_point(point: DrawPoint, heading: ElbowHeading, distance: f64) -> DrawPoint {
        DrawPoint::new(
            point.x + heading.dx() * distance,
            point.y + heading.dy() * distance,
        )
    }

    /// Total Manhattan path length across all segments.
    pub fn path_length(points: &[DrawPoint]) -> f64 {
        points
            .windows(2)
            .map(|segment| Self::manhattan_distance(segment[0], segment[1]))
            .sum()
    }

    /// Whether two point lists are element-wise equal.
    pub fn point_lists_equal(a: &[DrawPoint], b: &[DrawPoint]) -> bool {
        if a.len() != b.len() {
            return false;
        }

        a.iter().zip(b.iter()).all(|(left, right)| left == right)
    }

    /// Whether two point lists are equal except at the endpoints.
    pub fn point_lists_equal_except_endpoints(a: &[DrawPoint], b: &[DrawPoint]) -> bool {
        if a.len() != b.len() || a.len() < 2 {
            return false;
        }

        for index in 1..(a.len() - 1) {
            if a[index] != b[index] {
                return false;
            }
        }

        true
    }

    /// Resolves the bound heading for an arrow endpoint.
    ///
    /// Returns the cardinal direction the arrow should exit from the bound
    /// element, or `None` when the binding target is missing.
    pub fn resolve_bound_heading(
        binding: &ArrowBinding,
        elements_by_id: &HashMap<String, ArrowElementState>,
        point: DrawPoint,
    ) -> Option<ElbowHeading> {
        let element = elements_by_id.get(binding.element_id.as_str())?;
        let bounds = compute_element_world_aabb(element);
        let anchor = ArrowBindingUtils::resolve_elbow_anchor_point(binding, element);
        Some(Self::heading_for_point_on_bounds(
            bounds,
            anchor.unwrap_or(point),
        ))
    }
}

fn compute_element_world_aabb(element: &ArrowElementState) -> DrawRect {
    let rect = element.rect;
    let rotation = element.rotation;
    if rotation == 0.0 {
        return rect;
    }

    let center = rect.center();
    let half_width = rect.width().abs() / 2.0;
    let half_height = rect.height().abs() / 2.0;
    let cos_theta = rotation.cos().abs();
    let sin_theta = rotation.sin().abs();
    let x_extent = half_width * cos_theta + half_height * sin_theta;
    let y_extent = half_width * sin_theta + half_height * cos_theta;

    DrawRect::new(
        center.x - x_extent,
        center.y - y_extent,
        center.x + x_extent,
        center.y + y_extent,
    )
}
